### ISSUE TITLE
{.inline.} procs should be implicitly convertible to non-inline procs; faster closures

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#396]
there are 3 parts to this RFC, and they're related:
* allow implicit conversion from {.inline.} to a closure or nimcall convention
* generate a thunk for inline procs and also for nimcall procs: `proc fn(args)` => `proc fnWrap(args, ctx: pointer)`
* avoid branching when calling closures `f.ClE_0?` and just call the function pointer with the environment always

## proposal
inline procs should be implicitly convertible to non-line procs, in particular a proc accepting a closure proc argument should accept an inline proc:
```nim
proc call(f: proc(a: int)) = discard
proc f1(a: int) = discard
proc f2(a: int) {.inline.} = discard
call(f1) # already works
call(f2) # RFC: this now works
```

## assignment from non-inline to inline
Assigning a non-inline procvar to an inline proc would become legal too:
```nim
when true:
  var f: proc(a: int)
  proc f1(a: int) = discard
  proc f3(a: int) {.inline.} = discard
  proc main =
    var x = 0
    proc f2(a: int) {.closure.} = discard x
    f = f1
    f = f2
    # f = f3 # Error: type mismatch:
  main()
```
a subtle point is that the address of f and f3 may differ after f = f3 (implementation detail)

## more efficient closures for nimcall (the default calling convention)
A surprising benefit of this RFC is that this may result in more efficient closures:

on the following example, this RFC would eliminate the overhead of nimcall branching which, while predictable for a branch predictor, seems to cause problem for CPU instruction reordering
```
# without RFC
nim r -d:case5a -d:danger main # 0.26 seconds
nim r -d:case5b -d:danger main # 0.19 seconds (using {.inline.})

# with RFC
nim r -d:case5a -d:danger main # 0.19 seconds
nim r -d:case5b -d:danger main # 0.19 seconds (using {.inline.})
```

I manually implemented the effect of this RFC on this example by adding a thunk to convert `fn1` into a closure, see the tiny diff here: https://gist.github.com/timotheecour/962aa3712dc51099bf7cacc3f5e31ce8

the result of this is that we avoid the overhead of branching via `f.ClE_0?`, and it matters:

```nim
when true: # D20210701T210303
  when defined case5a: {.pragma: myprag.}
  when defined case5b: {.pragma: myprag, inline.}
  proc fn1(a: int, b: int): int {.myprag.} =
    result = (a shl 3) + (b shr 4)
    result += (a * result) - (a * b)
  proc bar(f: proc(a, b: int): int {.myprag.}): int =
    let n = 100_000_000
    for i in 0..<n:
      result += f(i, result)
  import std/times
  proc main =
    let t = cpuTime()
    let c = bar(fn1)
    let t2 = cpuTime()
    echo (t2 - t, c)
  main()
```

## note 1
* would fix https://github.com/nim-lang/Nim/issues/18220
* would enable improving performance of many APIs in stdlib by marking hotspots as `{.inline.}`, eg https://github.com/nim-lang/Nim/pull/18227 removed `{.inline.}` from `proc hash*[T: tuple | object | proc](x: T): Hash {.inline.}` and results in 1.6X slowdown (refs https://github.com/nim-lang/Nim/issues/18220#issuecomment-858410617)
* adding {.inline.} to some std/random APIs makes common code 10X faster, eg:
```nim
when defined case1: # D20210701T213234:here
  import std/random
  import std/times
  proc main() =
    var s = initRand()
    var c = 0
    let n = 100_000_000
    let t = cpuTime()
    for i in 0..<n:
      c+=rand(s, 1024)
    let t2 = cpuTime()
    echo (t2 - t, c)
  main()
```
with current nim (3ceaf5c1309ac8d4d68e6f39c13b021bcc1b15f4): 1.3s
with {.inline.} added to 3 APIS: 0.13s

## note 2
I tried a simplified version of this RFC as follows
```
const PatProc = "$1.ClE_0? $1.ClP_0($3$1.ClE_0):(($4)($1.ClP_0))($2)"
const PatProc = "$1.ClP_0($3$1.ClE_0)"
```
it essentially amounts to replacing:
```nim
if cl.ClE_0 != nil: cl.ClP_0(args, cl.ClE_0)
else: cast[f2](cl.ClP_0)(args)
```
by:
```nim
cl.ClP_0(args, cl.ClE_0)
```
which amounts, when the env is nil, to passing an extra argument on the stack (containing NULL); although it works on my system, IIUC it's UB according to C standard (casting a function type to another one and calling it is UB; it might depend on calling convention etc).

But this RFC doesn't have UB because it defines a wrapper thunk so that we can always call `cl.ClP_0(args, cl.ClE_0)`

**Comment by Araq** [comment]
Thorough analysis with one minor addition: An optimizer working on the assembler level could eliminate the conditional and the code bloat quite easily.

**Comment by juancarlospaco** [comment]
I tried adding LTO to the danger or release by default thinking kinda similar, but got closed. 🤷

**Comment by timotheecour** [comment]
> I tried adding LTO to the danger or release by default thinking kinda similar, but got closed. 🤷

it got closed for good reasons
* adding LTO guarantees a 4x increase in compile times
* there's no guarantee it increases RT performance, in fact compiling nim binary itself with LTO makes nim run slower
* it doesn't help with passing `{.inline.}` procs in contexts accepting non-inline procs; and if you just want to go ahead and remove all inline annotations and rely on LTO, that'd make all builds much slower if they don't have LTO, so you'd have to choose between 4X slower compile times or Nx slower runtimes (N could be very large for some inline procs, multiple orders of magnitude).

**Comment by juancarlospaco** [comment]
No, it was actually requested on a PR about adding `{.inline.}` on a tiny proc, I am Ok with `{.inline.}`.
So the compiler would transform the AST to remove the pragma or just ignore it?.
